Fix: Implement the get_asset_data function

wayzeek opened the pull request:
# Implement get_asset_data function

## Summary

Implements the `get_asset_data` function in the RWA Factory contract to return asset metadata for a given token ID.

## Changes

- **RWA Factory Contract**: Replaced unimplemented `get_asset_data` function with actual implementation that reads from storage
- **Interface**: Updated return type from `felt252` to `AssetData` struct and uncommented the import
- **Asset Struct**: Removed `Copy` trait from `AssetData` derive macro (not needed for storage)
- **Library**: Added structs module export to make `AssetData` accessible

## Technical Details

The function now reads asset data directly from the contract's storage using the existing `asset_data` storage variable. This provides a straightforward way to retrieve complete asset metadata by token ID.

Closes #77

wayzeek committed:
feat(asset): Introduce AssetData struct and update related interfaces

wayzeek committed:
Merge branch 'main' into implement-get-asset-data-function

GideonBature commented on src/lib.cairo:
```diff
@@ -10,4 +10,3 @@ pub mod events {
 pub mod structs {
```
Please exclude this.

GideonBature reviewed (commented):
Please sync your branch on github, pull from main, rebase main on your branch, resolve merge conflict, take care of the reviews and push your code. You are doing good so far. Thank you.

wayzeek committed:
Merge branch 'main' into implement-get-asset-data-function

wayzeek committed:
feat: Add tests for get_asset_data functionality and remove duplicate import

wayzeek commented:
@GideonBature I can not add direct tests for the `get_asset_data` function yet since the main RWAFactory contracts is not compiling.
Other tests are not testing their functions properly, for example in `test_tokenize_asset.cairo`:

```
#[test]
#[should_panic]
fn test_unauthorized_tokenize() {
    let (contract_instance, contract_address) = deploy_contract();
```

this passes because the `deploy_contract` function is panicking not the tokenize_asset function itself.
same for :

```
fn test_basic_functionality() {
    // Test that we can create and manipulate basic data structures
    let tokenizer = TOKENIZER();
    let owner = OWNER();
    let admin = ADMIN();

    // Test address constants work
    assert(tokenizer != owner, 'Addresses should be different');
    assert(admin != tokenizer, 'Admin be dff from tokenizer');
    assert(owner != admin, 'Owner should be dff from admin');

    // Test that we can create asset data
    let asset_data = AssetData {
        asset_type: 'PRECIOUS_METAL',
        name: "Gold Bar",
        description: "1kg gold bar",
        value_usd: 200,
        legal_doc_uri: "ipfs://gold",
        image_uri: "ipfs://gold-image",
        location: "Vault",
        created_at: 100,
    };

    assert(asset_data.asset_type == 'PRECIOUS_METAL', 'Asset type be precious metal');
    assert(asset_data.value_usd == 200, 'Value should be 200');
    assert(asset_data.name == "Gold Bar", 'Name should be Gold Bar');
}
```

this passes because he never calls the contract it is just reading from the struct.

my point is :
it is currently not possible to tests those methods since the main contract is not compiling (not even in lib.cairo), the other tests should not be passing as well, it is just mock data.

GideonBature commented:
> @GideonBature I can not add direct tests for the `get_asset_data` function yet since the main RWAFactory contracts is not compiling. Other tests are not testing their functions properly, for example in `test_tokenize_asset.cairo`:
> 
> ```
> #[test]
> #[should_panic]
> fn test_unauthorized_tokenize() {
>     let (contract_instance, contract_address) = deploy_contract();
> ```
> 
> this passes because the `deploy_contract` function is panicking not the tokenize_asset function itself. same for :
> 
> ```
> fn test_basic_functionality() {
>     // Test that we can create and manipulate basic data structures
>     let tokenizer = TOKENIZER();
>     let owner = OWNER();
>     let admin = ADMIN();
> 
>     // Test address constants work
>     assert(tokenizer != owner, 'Addresses should be different');
>     assert(admin != tokenizer, 'Admin be dff from tokenizer');
>     assert(owner != admin, 'Owner should be dff from admin');
> 
>     // Test that we can create asset data
>     let asset_data = AssetData {
>         asset_type: 'PRECIOUS_METAL',
>         name: "Gold Bar",
>         description: "1kg gold bar",
>         value_usd: 200,
>         legal_doc_uri: "ipfs://gold",
>         image_uri: "ipfs://gold-image",
>         location: "Vault",
>         created_at: 100,
>     };
> 
>     assert(asset_data.asset_type == 'PRECIOUS_METAL', 'Asset type be precious metal');
>     assert(asset_data.value_usd == 200, 'Value should be 200');
>     assert(asset_data.name == "Gold Bar", 'Name should be Gold Bar');
> }
> ```
> 
> this passes because he never calls the contract it is just reading from the struct.
> 
> my point is : it is currently not possible to tests those methods since the main contract is not compiling (not even in lib.cairo), the other tests should not be passing as well, it is just mock data.

So what is now your point ser? normally testing is for the functionality of whatsoever function/method you are righting, it is to test them to make sure whatsoever thing they are meant to do, they are doing it the way they're doing it... or am I missing anything?

GideonBature commented:
> @GideonBature I can not add direct tests for the `get_asset_data` function yet since the main RWAFactory contracts is not compiling. Other tests are not testing their functions properly, for example in `test_tokenize_asset.cairo`:
> 
> ```
> #[test]
> #[should_panic]
> fn test_unauthorized_tokenize() {
>     let (contract_instance, contract_address) = deploy_contract();
> ```
> 
> this passes because the `deploy_contract` function is panicking not the tokenize_asset function itself. same for :
> 
> ```
> fn test_basic_functionality() {
>     // Test that we can create and manipulate basic data structures
>     let tokenizer = TOKENIZER();
>     let owner = OWNER();
>     let admin = ADMIN();
> 
>     // Test address constants work
>     assert(tokenizer != owner, 'Addresses should be different');
>     assert(admin != tokenizer, 'Admin be dff from tokenizer');
>     assert(owner != admin, 'Owner should be dff from admin');
> 
>     // Test that we can create asset data
>     let asset_data = AssetData {
>         asset_type: 'PRECIOUS_METAL',
>         name: "Gold Bar",
>         description: "1kg gold bar",
>         value_usd: 200,
>         legal_doc_uri: "ipfs://gold",
>         image_uri: "ipfs://gold-image",
>         location: "Vault",
>         created_at: 100,
>     };
> 
>     assert(asset_data.asset_type == 'PRECIOUS_METAL', 'Asset type be precious metal');
>     assert(asset_data.value_usd == 200, 'Value should be 200');
>     assert(asset_data.name == "Gold Bar", 'Name should be Gold Bar');
> }
> ```
> 
> this passes because he never calls the contract it is just reading from the struct.
> 
> my point is : it is currently not possible to tests those methods since the main contract is not compiling (not even in lib.cairo), the other tests should not be passing as well, it is just mock data.

Just like I said, follow what is done [here](https://github.com/starknetprojects/rwax/blob/main/tests/test_tokenize_asset.cairo). This is testing the contract. The other test was just merely testing the functions. Now, just focus on yours and do the right thing by testing the contract as you should. Thank you. Just like I said, look at the current code from the main repository, and not the one you fork, you can also sync your fork to be on the same page with the upstream.

wayzeek commented:
you are right the thing is is that in order to make tests on a contract, the contract must compile.
at the current state it is not compiling so we can not directly tests the methods on this contract.

no compiling = no deploying = no testing

i would recommend adding :
```
pub mod contracts {
    pub mod rwa_factory;
}
```
to `src/lib.cairo`

you'll see there are a lot of errors in the contracts, preventing it from compiling.
again if it does not compile, we cant deploy it and we cant call the methods on the contract.

GideonBature commented:
Other test is not supposed to disturb your own test brother, because you are writing a test in another file. Also, the project is compiling, so I don't really understand what you mean by the contract is not compiling.

wayzeek committed:
fix: fixed formatting

GideonBature commented:
> check your `target/` folder, you won't see your compiled contract in there :) it is not compiling it since it is not included in the `lib.cairo`. Check this out : https://www.starknet.io/cairo-book/ch01-02-hello-world.html#building-a-scarb-project
> 
> if your contract was being compiled you would have a : `rwa_factory.sierra.json` file in `target/dev`

You don't need that to test brother. We are not deploying yet boss. And for that, we only need to enable it in the Scarb.toml file, which for now is not necessary. So if possible mock a test using `snforge`. Then proceed to write test and all of that...

wayzeek commented:
no worries, you will see quite a lot of errors, mostly coming from missing imports.

what i would recommend for you is :
1. Make a new issue to fix compilation errors on the contract
2. Once this passes, make a new issue to update all tests to not use mock data but instead use the contract methods directly (now we can since it will be compiling)

GideonBature commented:
> no worries, you will see quite a lot of errors, mostly coming from missing imports.
> 
> what i would recommend for you is :
> 
> 1. Make a new issue to fix compilation errors on the contract
> 2. Once this passes, make a new issue to update all tests to not use mock data but instead use the contract methods directly (now we can since it will be compiling)

Thank you, was having issues with my cairo1 extension and scarb, had to restart my PC this morning to be able to fix the errors. So I think you can go ahead wit h your own implementation and test. Thank you for your insight. I appreciate.

wayzeek committed:
Merge branch 'main' into implement-get-asset-data-function

wayzeek commented:
good job @GideonBature , I will add the tests now :)
For the CI/CL pipeline make sure you check :
`scarb build`
`scarb fmt --check`
`scarb test` :)

GideonBature commented on tests/test_get_asset_data.cairo:
```diff
@@ -0,0 +1,78 @@
+use rwax::structs::asset::AssetData;
```
I know that you complained of the smart contract not compiling properly, now it does, so I believe you should be able to write the test now. correct?

wayzeek committed:
Merge remote-tracking branch 'origin/main' into implement-get-asset-data-function

wayzeek committed:
Enhance asset data tests with tokenization scenarios

- Added deployment and tokenization helper functions for testing.
- Updated tests to verify asset data retrieval after tokenization.
- Included tests for multiple tokenized assets and handling of nonexistent tokens.
- Improved asset data structure with new test cases for various asset types.

wayzeek commented:
Hello @GideonBature , this is done :)

GideonBature commented:
@wayzeek seems you may need to fix your formatting.

wayzeek committed:
fix: formatting

wayzeek commented:
done :) @GideonBature

GideonBature reviewed (approved):
ACK [54c1f6f](https://github.com/starknetprojects/rwax/pull/90/commits/54c1f6f367e939ef8eb09889ce2050756b5c68bd)
